### ISSUE TITLE
Fix FP S4782 (`no-redundant-optional`): Do not raise issues when "exactOptionalPropertyTypes" is enabled

### ISSUE THREAD

**Issue by yassin-kammoun-sonarsource** [SonarSource/SonarJS#4415]
The rule raises false-positive issues when TypeScript's compiler option [exactOptionalPropertyTypes](https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes) is enabled:

```typescript
// Assuming `compilerOptions: { "exactOptionalPropertyTypes": true }`

interface ConnData {
  promise?: Promise<unknown> | undefined; // False-positive: `?` is not redundant
}
```

Instead, the rule should retrieve the compiler options from TypeScript compiler API and turn the rule into a noop if the option is enabled.

https://community.sonarsource.com/t/typescript-linting-optional-vs-undefined/104802

**Comment by Wohops** [comment]
Migrated to Jira. New ticket number: JS-67
